### ISSUE TITLE
OSX Shortcuts don't work and/or don't exist.

### ISSUE THREAD

**Issue by autismuk** [pkulchenko/ZeroBraneStudio#298]
OSX keyboards do not necessarily have Page up and Page Down keys. 
Ctrl+TAB does not work. To be precise it sometimes works, it sometimes switches to the second file and  sticks there even though there are four files open. Sometimes it switches the second file of many and sometimes it opens the Locale Console.  I think whatever is dispatching key events is majorly broken here, it seems to handle the tab switch pretty much randomly. This may be because I have a UK small keyboard, I don't know.

Would suggest using CMD 1/2/3/4 perhaps for selecting tabs ?

The shortcuts are wrong. It says fn F5 stops , but it's actually shift + F5. (It would also be nice to restart on F5 if it was already running.)

**Comment by pkulchenko** [comment]
`Ctrl-(Shift-)Tab` is not working on OSX (which appears to be a wxwidgets but) as described in [this discussion](http://wxwidgets.10942.n7.nabble.com/Ctrl-Shift-Tab-doesn-t-properly-navigate-over-tabs-on-OSX-wxwidgets3-1-td86834.html).

**Comment by IonoclastBrigham** [comment]
So this is actually a few different issues that look related, but aren't.

As of latest commit on github (7a8027e):

* `Ctrl-TAB` and related shortcuts are still wonky. What's happening is it's selecting other tabs in the ZBS UI, besides just the open files. The order of cycling through your open files also seems a little odd, and sometime certain ones still can't be reached. **Test case:** single-click any file or directory in the "Project" tab. You can now control-tab to "Outline"; it will look like nothing is happening for a few presses, then it will cycle through all your open files, and eventually the output and console tabs, where it gets stuck.
* Menu items mapped to `FX` keys still show the `fn` prefix but not the modifier, if you use the function keys as their media key equivalents by default, despite being more than a year since that patch to WX widgets. Not sure what happened here. It may have been re-introduced at their end, the patch may have been rejected or reverted for some reason, or it was maybe never imported properly into ZBS somehow.

And while we're talking keyboard shortcut bugs on OS X, I'll throw in one of my own:

* The docs say, "On OSX, the Command key is mapped to wxSTC_SCMOD_CTRL and the Control key to wxSTC_SCMOD_META." This echoes the equivalent documentation for Scintilla. Basically everything that uses `Ctrl` does in fact map to `Cmd`, but I haven't been able to get any shortcut specified with `wxSTC_SCMOD_META` to work at all with my physical `Ctrl` key. It just inserts a little token marking the ASCII control code that it maps to in a Unix terminal, e.g. `Ctrl-A` inserts a `SOH` token (ASCII code 1).

This is important to me because I'm used to typical OS X style key bindings and Emacs key bindings. I have little to no interest in learning *yet another* keymap configuration. Consider:

    editor.keymap[#editor.keymap+1] = {('A'):byte(), wxstc.wxSTC_SCMOD_META, wxstc.wxSTC_CMD_HOME, "Macintosh"}

This should actually be `Ctrl-A` (or `C-a` in Emacs parlance), but of course, using `wxstc.wxSTC_SCMOD_CTRL` would map this to the `Cmd` key on a Mac (which is already bound to "select all"). So, following the docs, I set the modifier to `Meta`, which should map to the physical `Ctrl` key, but it just isn't recognized.

Any other single or combination of modifiers is recognized correctly, e.g. all my `Alt` shortcuts work. And `Ctrl-TAB` works (more or less), so I'm not sure how that's made to happen.

**Comment by IonoclastBrigham** [comment]
Okay, I'm seeing in `src/editor/editor.lua`, lines 1205 - 1214, it looks for the the tab-switching shortcuts using `wx.wxMOD_RAW_CONTROL` which (I presume) always maps to the physical `Ctrl` keys, regardless of platform.

At runitme, it's value is 16, which is the same value as `wxstc.wxSTC_SCMOD_META` on a Mac. So that all looks legit. But, for whatever reason, using this value in an `editor.keymap` shortcut doesn't work.

**Comment by pkulchenko** [comment]
@IonoclastBrigham, yes, you are correct on these points;

> Ctrl-TAB and related shortcuts are still wonky. 

Yes, this is a wxwidgets issue, as described in the [discussion I linked in the previous comment](http://wxwidgets.10942.n7.nabble.com/Ctrl-Shift-Tab-doesn-t-properly-navigate-over-tabs-on-OSX-wxwidgets3-1-td86834.html); I couldn't find any workaround to make it work and it's not yet fixed in the trunk version of wxwidgets.

> Menu items mapped to FX keys still show the fn prefix but not the modifier

Correct; this is described earlier in this thread as well and will be fixed by #260.

> At runitme, it's value is 16, which is the same value as wxstc.wxSTC_SCMOD_META on a Mac. So that all looks legit. But, for whatever reason, using this value in an editor.keymap shortcut doesn't work. 

The description looks correct; I'll check on why it may be not working.

**Comment by IonoclastBrigham** [comment]
Oh, indeed. I didn't realize the fix for the F-keys was in a new major revision of wx. That's reasonable this isn't fixed in ZBS yet, then.

Let me know if there's anything I can do to help with `Ctrl` not mapping correctly. I don't mind rummaging through the source (obviously) but I don't quite know it well enough to track this one down so far.

For reference, here's my user.lua: https://gist.github.com/IonoclastBrigham/23cf2aa494de3259deb3 You'll notice a lot of debugging junk I imported to try to figure out what's going on, and some metatable stuff I was playing with at the top. I'll point out that all the `Cmd` (spelled `CTRL`) and `Alt` shortcuts work, but none of the `Ctrl` (spelled `META`) shortcuts do.

**Comment by pkulchenko** [comment]
I checked Scintilla code in wxwidgets and none of their "default" shortcuts that involve `Ctrl` key doesn't work either. For example, `Ctrl-Right` should jump to the next word and `Shift-Ctrl-Right` should extend the selection to the next word, but nothing happens when I try it.

I'll send an email to Scintilla maillist to check, but it doesn't seem like it's something I can fix in ZBS.

**Comment by IonoclastBrigham** [comment]
If you gain any insight as to where to look but not a workaround or fix from them, I might go poke around in the wx or sci source, as appropriate, see if I can fix up a patch.

**Comment by pkulchenko** [comment]
@IonoclastBrigham, I found this ticket in wxwidgets that may be directly related to the issue we see: http://trac.wxwidgets.org/ticket/16243; if that's the case, the issue is in wxwidgets and not Scintilla. I haven't had a chance to try the patch yet; if you want to give it a try, you'll need to tweak `build/build-macosx.sh` script to use an existing (patched) wxwidgets install instead of extracting a new one. The easiest way is to let it extract and compile it (to make sure everything works as expected) and then patch and comment out the extraction commands in the script.

The command that will build new version of wxlua library is `(cd build; ./build-macosx.sh lua wxwidgets wxlua)`. You can run it with `wxwidgets` first, just make sure you comment out the line that deletes the files after compilation.

**Comment by pkulchenko** [comment]
There is also a related ticket: http://trac.wxwidgets.org/ticket/16219. It's a proposed patch that was recommended to submit upstream (to Scintilla authors). I'm not sure if it was done as I don't see the changes in the current source of Scintilla. This would only affect the "default" set of bindings and having the patch I referenced earlier would allow applications like ZBS to set news mappings, so it would be better for our case.

**Comment by IonoclastBrigham** [comment]
Okay, so 16219 was closed as won't-fix, punting the patch upstream for integration into Scintilla. But the person who filed 16243 said he already tried to submit *his* patch to upstream and it was rejected because the Scintilla devs view wx as a homogeneous platform in its own right, regardless of the underlying OS.

To me, this sounds like neither of them want to budge, and it's unlikely to get resolved. Ever. Unless one or both projects got a big enough donation from a major corporation or something, to pressure them to fix it. The first patch may be a viable solution from a technical stand point, but will probably be rejected by wx for the same reason: they don't want to be out of sync with upstream Scintilla.

I'm not sure where all the keyboard input processing happens throughout the ZBS codebase, but I do know for a fact that detecting the `Ctrl` key does work, since it's handled for tab switching (more on that later). As a stop-gap, we might want to handle any user-mapped use of `Ctrl` shortcuts in a similar way, special-casing that event when we detect we're on OS X. Maybe decorate it with `FIXME` comments to mark it as tech-debt, and add a pending issue to the ZBS tracker, for if/when one of our upstream projects does something.

**Comment by IonoclastBrigham** [comment]
And regarding the tab switching bug: I just found that with the latest from master, `Ctrl-Tab` is very flaky and unreliable as described, but `Ctrl-PgUp/PgDn` *works every time!* It visits all your open tabs, in order, and even wraps around. This kinda makes it seem like the issue is either in the underlying library detecting or reporting specifically the `Ctrl-Tab` pair (or perhaps `Tab` itself), or else some subtlety in the handler in `src/editor/editor.lua` that makes it respond inconsistently. I have no idea what kind of bug that would be, but I'm going to investigate a little more in a moment.

**Comment by pkulchenko** [comment]
> The first patch may be a viable solution from a technical stand point, but will probably be rejected by wx for the same reason: they don't want to be out of sync with upstream Scintilla.

It may or may not be as it doesn't deal with any of the defines and it's in the Scintilla code controlled by wxwidgets (and the other patch was in the Scintilla code itself, so it was rightfully rejected as it's too brittle to keep the local changes there).

> I just found that with the latest from master, Ctrl-Tab is very flaky and unreliable as described, but Ctrl-PgUp/PgDn works every time! 

Right, this is because I handle `Ctrl-PgUp/PgDn` in ZBS, but the same code that is *supposed* to handle `Ctrl-Tab` simply doesn't get any of those events, so the logic doesn't apply there. You can see this easily by adding `DisplayOutputLn` call to the handler that proceses `Ctrl-PgUp/PgDn` or `Ctrl-Tab` and you'll see that none of the Tab-related events get there. I couldn't find *any* way to make `Ctrl-Tab` work on OSX (as described [here](http://wxwidgets.10942.n7.nabble.com/Ctrl-Shift-Tab-doesn-t-properly-navigate-over-tabs-on-OSX-wxwidgets3-1-td86834.html)).

**Comment by pkulchenko** [comment]
> Maybe decorate it with FIXME comments to mark it as tech-debt, and add a pending issue to the ZBS tracker, for if/when one of our upstream projects does something.

As the last resort, I can always patch wxwidgets locally before compiling, but I'd very much prefer not doing this for obvious reasons. Having said that, I do patch wxlua sources, so I'm not strongly against that, but would prefer this to be fixed in wxwidgets.

**Comment by pkulchenko** [comment]
@IonoclastBrigham, I've tested the patch, but it didn't work for me without couple of changes. See the [comment in the ticket](http://trac.wxwidgets.org/ticket/16243#comment:3) for details.

**Comment by IonoclastBrigham** [comment]
Might be reasonable to have an "apply any patches from this directory" step
after extracting source, in the build script. This lets us reference a true
vanilla upstream, and still get this and any future fixes in there. Many
package managers do something similar (e.g. macports).

Or we could just do it by hand, if you think it'll get accepted into wx
soon.

On Thu, May 21, 2015 at 12:05 PM, Paul Kulchenko <notifications@github.com>
wrote:

> @IonoclastBrigham <https://github.com/IonoclastBrigham>, I've tested the
> patch, but it didn't work for me without couple of changes. See the comment
> in the ticket <http://trac.wxwidgets.org/ticket/16243#comment:3> for
> details.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pkulchenko/ZeroBraneStudio/issues/298#issuecomment-104390735>
> .
>



-- 
Brigham Toskin
Ionoclast Laboratories, LLC

**Comment by pkulchenko** [comment]
@IonoclastBrigham, @autismuk, I pushed a workaround for the issue with `Ctrl-Tab` navigation on OSX. It should make it work in most cases, but since it's a workaround, I'd be interested to know if you notice anything not working as expected because of it (especially something that involves `Ctrl` key).

The main issue is described in [this wxwidgets ticket](http://trac.wxwidgets.org/ticket/17064), and since I don't have a solution for it, I found a workaround that worked well in my testing.

The rest of the issues (the menu labels inconsistencies) will be addressed by the upgrade ticket (#260).
